include/alpaka/event/EventCpu.hpp

```diff
@@ -534,14 +534,13 @@ namespace alpaka
                     // Copy the shared pointer of the event implementation.
                     // This is forwarded to the lambda that is enqueued into the stream to ensure that the event implementation is alive as long as it is enqueued.
                     auto spEventCpuImpl(event.m_spEventCpuImpl);
-
+                    std::lock_guard<std::mutex> lk(spEventCpuImpl->m_Mutex);
+                    if(!spEventCpuImpl->m_bIsReady)
                     {
-                        std::lock_guard<std::mutex> lk(spEventCpuImpl->m_Mutex);
                         spEventCpuImpl->m_bIsWaitedFor = true;
+                        // NOTE: Difference to async version: directly wait for event.
+                        wait::wait(spEventCpuImpl);
```

Comment:
I am not sure if the synchronous case will work on all platforms.
This `WaiterWaitFor` specialization locks the mutex and directly calls `wait::wait(...)` on the event. This in turn calls `CurrentThreadWaitFor` (line 443) which tries to lock the mutex again leading to a deadlock. This would require a recursive mutex which is not used currently.

I would say that we can delete lines 534-540 as well as line 543 and keep only this line and its comment. This would solve the recursive mutex call. And remove duplicate code that is executed in the  `wait::wait(...)` -> `CurrentThreadWaitFor` anyhow.